Concurrency issue in the end to end tests

The build [\#1150](https://travis-ci.org/apache/camel-k/builds/559896454) failed with:

```
=== RUN   TestPlatformUpgrade
Camel K installed in namespace test-7ae995dd-f1fe-4a9f-b243-99b98ffaf955
--- FAIL: TestPlatformUpgrade (7.94s)
    upgrade_test.go:40: 
        Expected
            <*errors.StatusError | 0xc000434240>: {
                ErrStatus: {
                    TypeMeta: {Kind: "", APIVersion: ""},
                    ListMeta: {SelfLink: "", ResourceVersion: "", Continue: ""},
                    Status: "Failure",
                    Message: "Operation cannot be fulfilled on deployments.apps \"camel-k-operator\": the object has been modified; please apply your changes to the latest version and try again",
                    Reason: "Conflict",
                    Details: {
                        Name: "camel-k-operator",
                        Group: "apps",
                        Kind: "deployments",
                        UID: "",
                        Causes: nil,
                        RetryAfterSeconds: 0,
                    },
                    Code: 409,
                },
            }
        to be nil
```

Looks like a concurrency issue in the end to end test suite or in the operator.

--- a/e2e/test_support.go
+++ b/e2e/test_support.go
@@ -27,6 +27,8 @@ import (
 	"fmt"
 	"time"
 
+	"io/ioutil"
+
 	"github.com/apache/camel-k/pkg/apis/camel/v1alpha1"
 	"github.com/apache/camel-k/pkg/client"
 	"github.com/apache/camel-k/pkg/cmd"
@@ -36,7 +38,6 @@ import (
 	"github.com/onsi/gomega"
 	projectv1 "github.com/openshift/api/project/v1"
 	"github.com/spf13/cobra"
-	"io/ioutil"
 	appsv1 "k8s.io/api/apps/v1"
 	v1 "k8s.io/api/core/v1"
 	k8serrors "k8s.io/apimachinery/pkg/api/errors"
@@ -369,40 +370,42 @@ func operatorTryPodForceKill(ns string) {
 	}
 }
 
-func scaleOperator(ns string, replicas int32) error {
-	lst := appsv1.DeploymentList{
-		TypeMeta: metav1.TypeMeta{
-			Kind:       "Deployment",
-			APIVersion: appsv1.SchemeGroupVersion.String(),
-		},
-	}
-	opts := k8sclient.ListOptions{
-		LabelSelector: labels.SelectorFromSet(labels.Set{
-			"camel.apache.org/component": "operator",
-		}),
-		Namespace: ns,
-	}
-	if err := testClient.List(testContext, &opts, &lst); err != nil {
-		return err
-	}
-	if len(lst.Items) == 0 {
-		return errors.New("camel k operator not found")
-	} else if len(lst.Items) > 1 {
-		return errors.New("too many camel k operators")
-	}
+func scaleOperator(ns string, replicas int32) func() error {
+	return func() error {
+		lst := appsv1.DeploymentList{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Deployment",
+				APIVersion: appsv1.SchemeGroupVersion.String(),
+			},
+		}
+		opts := k8sclient.ListOptions{
+			LabelSelector: labels.SelectorFromSet(labels.Set{
+				"camel.apache.org/component": "operator",
+			}),
+			Namespace: ns,
+		}
+		if err := testClient.List(testContext, &opts, &lst); err != nil {
+			return err
+		}
+		if len(lst.Items) == 0 {
+			return errors.New("camel k operator not found")
+		} else if len(lst.Items) > 1 {
+			return errors.New("too many camel k operators")
+		}
 
-	operatorDeployment := lst.Items[0]
-	operatorDeployment.Spec.Replicas = &replicas
-	err := testClient.Update(testContext, &operatorDeployment)
-	if err != nil {
-		return err
-	}
+		operatorDeployment := lst.Items[0]
+		operatorDeployment.Spec.Replicas = &replicas
+		err := testClient.Update(testContext, &operatorDeployment)
+		if err != nil {
+			return err
+		}
 
-	if replicas == 0 {
-		// speedup scale down by killing the pod
-		operatorTryPodForceKill(ns)
+		if replicas == 0 {
+			// speedup scale down by killing the pod
+			operatorTryPodForceKill(ns)
+		}
+		return nil
 	}
-	return nil
 }
 
 /*
@@ -462,7 +465,7 @@ func deleteTestNamespace(ns metav1.Object) {
 	}
 
 	// Wait for all pods to be deleted
-	gomega.Eventually(numPods(ns.GetName()), 30 * time.Second).Should(gomega.Equal(0))
+	gomega.Eventually(numPods(ns.GetName()), 30*time.Second).Should(gomega.Equal(0))
 }
 
 func newTestNamespace() metav1.Object {

--- a/e2e/upgrade_test.go
+++ b/e2e/upgrade_test.go
@@ -37,7 +37,7 @@ func TestPlatformUpgrade(t *testing.T) {
 		Eventually(platformVersion(ns)).Should(Equal(defaults.Version))
 
 		// Scale the operator down to zero
-		Expect(scaleOperator(ns, 0)).Should(BeNil())
+		Eventually(scaleOperator(ns, 0), 10*time.Second).Should(BeNil())
 		Eventually(operatorPod(ns)).Should(BeNil())
 
 		// Change the version to an older one
@@ -45,7 +45,7 @@ func TestPlatformUpgrade(t *testing.T) {
 		Eventually(platformVersion(ns)).Should(Equal("an.older.one"))
 
 		// Scale the operator up
-		Expect(scaleOperator(ns, 1)).Should(BeNil())
+		Eventually(scaleOperator(ns, 1)).Should(BeNil())
 		Eventually(operatorPod(ns)).ShouldNot(BeNil())
 
 		// Check the platform version change
@@ -65,7 +65,7 @@ func TestIntegrationUpgrade(t *testing.T) {
 		initialImage := integrationPodImage(ns, "js")()
 
 		// Scale the operator down to zero
-		Expect(scaleOperator(ns, 0)).Should(BeNil())
+		Eventually(scaleOperator(ns, 0)).Should(BeNil())
 		Eventually(operatorPod(ns)).Should(BeNil())
 
 		// Change the version to an older one
@@ -76,7 +76,7 @@ func TestIntegrationUpgrade(t *testing.T) {
 		Eventually(kitsWithVersion(ns, defaults.Version)).Should(Equal(0))
 
 		// Scale the operator up
-		Expect(scaleOperator(ns, 1)).Should(BeNil())
+		Eventually(scaleOperator(ns, 1)).Should(BeNil())
 		Eventually(operatorPod(ns)).ShouldNot(BeNil())
 		Eventually(operatorPodPhase(ns)).Should(Equal(v1.PodRunning))
 

